Source/NSubstitute/Core/PendingSpecification.cs

```diff
@@ -1,29 +1,65 @@
+using System;
+
 namespace NSubstitute.Core
 {
     public class PendingSpecification : IPendingSpecification
     {
-        ICallSpecification _pendingSpec;
+        private readonly ISubstitutionContext _substitutionContext;
+
+        public PendingSpecification(ISubstitutionContext substitutionContext)
+        {
+            _substitutionContext = substitutionContext;
+        }
+
+        public bool HasPendingCallSpecInfo()
+        {
+            //Check ownership to avoid potential issues when spec from foreign substitution is present.
+            var currentInfo = _substitutionContext.ThreadLocalPendingSpecificationInfo as PendingSpecificationInfo;
+            return currentInfo != null && currentInfo.Owner == this;
+        }
 
-        public bool HasPendingCallSpec()
+        public IPendingSpecificationInfo UseCallSpecInfo()
         {
-            return _pendingSpec != null;
+            var info = _substitutionContext.ThreadLocalPendingSpecificationInfo as PendingSpecificationInfo;
+
+            if (info != null && info.Owner != this)
+            {
+                //That should never happen. It could only happen if there are some bugs or unexpected usages.
+                throw new InvalidOperationException("Specification from foreign substitution is present on stack.");
+            }
+
+            Clear();
+
+            return info;
         }
 
-        public ICallSpecification UseCallSpec()
+        public void SetCallSpecification(ICallSpecification callSpecification)
         {
-            var specToUse = _pendingSpec;
-            _pendingSpec = null;
-            return specToUse;
+            _substitutionContext.ThreadLocalPendingSpecificationInfo = new PendingSpecificationInfo(callSpecification, null, this);
+        }
+
+        public void SetLastCall(ICall lastCall)
+        {
+            _substitutionContext.ThreadLocalPendingSpecificationInfo = new PendingSpecificationInfo(null, lastCall, this);
         }
 
         public void Clear()
         {
-            UseCallSpec();
+            _substitutionContext.ThreadLocalPendingSpecificationInfo = null;
         }
 
-        public void Set(ICallSpecification callSpecification)
+        private class PendingSpecificationInfo: IPendingSpecificationInfo
         {
-            _pendingSpec = callSpecification;
+            public ICallSpecification CallSpecification { get; }
+            public ICall LastCall { get; }
+            public IPendingSpecification Owner { get; }
+
+            public PendingSpecificationInfo(ICallSpecification callSpecification, ICall lastCall, IPendingSpecification owner)
```

Comment:
The reason why I did that is because I'm uncertain in this code. That should never happen, but we are working with thread static scope and it could happen that we overlooked something.

In fact we can remove that checks at all - I see that it's quite impossible to have situation when last router points to one substitution, while thread local pending specs are from other substitution. But if it's possible to happen, we'll have a bug reported and could investigate that. Otherwise, we could just guess why sometimes behavior is strange.

If you feel that it's better to remove that check at all and assume that it should never happen - let me know and I'll remove it.